src/webgpu/api/validation/image_copy/buffer_related.spec.ts

```diff
@@ -0,0 +1,149 @@
+        GPUConst.BufferUsage.COPY_SRC | GPUConst.BufferUsage.COPY_DST,
+      ])
+  )
+  .fn(async t => {
+    const { method, usage } = t.params;
+
+    const buffer = t.device.createBuffer({
+      size: 16,
+      usage,
+    });
+
+    const success =
+      method === 'CopyB2T'
+        ? (usage & GPUBufferUsage.COPY_SRC) !== 0
+        : (usage & GPUBufferUsage.COPY_DST) !== 0;
+
+    const texture = t.device.createTexture({
+      size: { width: 2, height: 2, depthOrArrayLayers: 1 },
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    t.testBuffer(
+      buffer,
+      texture,
+      { bytesPerRow: 0 },
+      { width: 0, height: 0, depthOrArrayLayers: 0 },
+      { dataSize: 16, method, success }
+    );
+  });
+
+g.test('bytes_per_row_alignment')
+  .desc(
+    `Test that bytesPerRow must be a multiple of 256 for CopyB2T and CopyT2B if it is required.`
+  )
+  .params(u =>
+    u //
+      .combine('method', kImageCopyTypes)
+      .combine('dimension', ['2d', '3d'] as const)
+      .combine('format', kSizedTextureFormats)
+      .filter(({ dimension, format }) => textureDimensionAndFormatCompatible(dimension, format))
+      .filter(formatCopyableWithMethod)
+      .beginSubcases()
+      .combine('bytesPerRow', [undefined, 0, 1, 255, 256, 257, 512])
+      .combine('copyHeight', [0, 1, 2, 4, 8])
+      // copyHeight must be aligned with blockHeight
+      .filter(({ format, copyHeight }) => copyHeight % kTextureFormatInfo[format].blockHeight === 0)
+      // bytesPerRow must be specified and it must be equal or greater than the bytes size of each row if we are copying multiple rows.
+      // Note that we are copying one single block on each row in this test.
+      .filter(
+        ({ format, bytesPerRow, copyHeight }) =>
+          (bytesPerRow === undefined && copyHeight / kTextureFormatInfo[format].blockHeight <= 1) ||
+          (bytesPerRow !== undefined && bytesPerRow >= kTextureFormatInfo[format].bytesPerBlock)
+      )
+  )
+  .fn(async t => {
+    const { method, dimension, format, bytesPerRow, copyHeight } = t.params;
+
+    const info = kTextureFormatInfo[format];
+    await t.selectDeviceOrSkipTestCase(info.feature);
+
+    const buffer = t.device.createBuffer({
+      size: 512 * 8 * 16,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    let success = false;
+    // writeTexture doesn't require bytesPerRow to be 256-byte aligned.
+    if (method === 'WriteTexture') success = true;
+    // If the copy height <= 1, bytesPerRow is not required.
+    if (copyHeight / info.blockHeight <= 1 && bytesPerRow === undefined) success = true;
+    // If bytesPerRow > 0 and it is a multiple of 256, it will succeeed if other parameters are valid.
+    if (bytesPerRow !== undefined && bytesPerRow > 0 && bytesPerRow % 256 === 0) success = true;
+
+    const size = [info.blockWidth, !copyHeight ? info.blockHeight : copyHeight, 1];
```

Comment:
nit: I think this is equivalent? and clearer; I initially wasn't sure whether copyHeight could be undefined or something.
```suggestion
    const size = [info.blockWidth, copyHeight === 0 ? info.blockHeight : copyHeight, 1];
```

Reply:
Done.